File: ruby/zork/v1/game_map.rb
require "./room"

class GameMap
  attr_accessor :starting_room

  def initialize
    seed_room = Room.new(name: "Ye Olde Tavern", description: "There's nothing here but ale.")
    test_room = Room.new(name: "The Back Room", description: "This is another room.", rooms: { :south => seed_room })
    seed_room.rooms = { :north => test_room }
    @starting_room = seed_room
  end

  def get_opposite_direction direction
    { :north => :south, :east => :west, :south => :north, :west => :east }[direction]
  end

  def generate_unexplored_room previous_room, direction
    name = ("a".."z").to_a.sample(15).join
    opposite_direction = get_opposite_direction(direction)
    new_room = Room.new(name: name, rooms: { opposite_direction => previous_room })
    previous_room.rooms[direction] = new_room
  end
end

File: lua/unsorted/zork/game_map.lua
local utils = require "utils"
local Room = require "room"

GameMap = {}
function GameMap:new(starting_room)
   return Class.new(self, { starting_room = starting_room })
end

function GameMap:get_opposing_direction(direction)
   return ({
      NORTH = "SOUTH",
      EAST = "WEST",
      SOUTH = "NORTH",
      WEST = "EAST"
   })[direction]
end

function GameMap:generate_unexplored_room(previous_room, direction)
   local name = utils.join("", utils.pick(utils.chars("abcdefghijklmnopqrstuvwxyz"), 15))
   local opposite_direction = self:get_opposing_direction(direction)
   local new_room = Room:new(name, nil, { [opposite_direction] = previous_room })
   previous_room.rooms[direction] = new_room
   return new_room
end

return GameMap
